No instances available for uaa when start new microservice application

##### **Overview of the issue**

~~~
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'tokenStore' defined in class path resource [it/seiton/turnigo/web/mobgateway/config/MicroserviceSecurityConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.oauth2.provider.token.TokenStore]: Factory method 'tokenStore' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jwtAccessTokenConverter' defined in class path resource [it/seiton/turnigo/web/mobgateway/config/MicroserviceSecurityConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.oauth2.provider.token.store.JwtAccessTokenConverter]: Factory method 'jwtAccessTokenConverter' threw exception; nested exception is java.lang.IllegalStateException: No instances available for uaa
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:599)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1123)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1018)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:510)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
        at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
        at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:207)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1214)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1054)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1019)
        at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:566)
        ... 22 common frames omitted
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.oauth2.provider.token.TokenStore]: Factory method 'tokenStore' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jwtAccessTokenConverter' defined in class path resource [it/seiton/turnigo/web/mobgateway/config/MicroserviceSecurityConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.oauth2.provider.token.store.JwtAccessTokenConverter]: Factory method 'jwtAccessTokenConverter' threw exception; nested exception is java.lang.IllegalStateException: No instances available for uaa
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:189)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:588)
        ... 35 common frames omitted
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jwtAccessTokenConverter' defined in class path resource [it/seiton/turnigo/web/mobgateway/config/MicroserviceSecurityConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.oauth2.provider.token.store.JwtAccessTokenConverter]: Factory method 'jwtAccessTokenConverter' threw exception; nested exception is java.lang.IllegalStateException: No instances available for uaa
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:599)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1123)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1018)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:510)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
        at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
        at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:381)
        at it.seiton.turnigo.web.mobgateway.config.MicroserviceSecurityConfiguration$$EnhancerBySpringCGLIB$$8a3a3561.jwtAccessTokenConverter(<generated>)
        at it.seiton.turnigo.web.mobgateway.config.MicroserviceSecurityConfiguration.tokenStore(MicroserviceSecurityConfiguration.java:60)
        at it.seiton.turnigo.web.mobgateway.config.MicroserviceSecurityConfiguration$$EnhancerBySpringCGLIB$$8a3a3561.CGLIB$tokenStore$1(<generated>)
        at it.seiton.turnigo.web.mobgateway.config.MicroserviceSecurityConfiguration$$EnhancerBySpringCGLIB$$8a3a3561$$FastClassBySpringCGLIB$$5c61a0cd.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:228)
        at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:356)
        at it.seiton.turnigo.web.mobgateway.config.MicroserviceSecurityConfiguration$$EnhancerBySpringCGLIB$$8a3a3561.tokenStore(<generated>)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:162)
        ... 36 common frames omitted
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.oauth2.provider.token.store.JwtAccessTokenConverter]: Factory method 'jwtAccessTokenConverter' threw exception; nested exception is java.lang.IllegalStateException: No instances available for uaa
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:189)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:588)
        ... 57 common frames omitted
Caused by: java.lang.IllegalStateException: No instances available for uaa
        at org.springframework.cloud.netflix.ribbon.RibbonLoadBalancerClient.execute(RibbonLoadBalancerClient.java:79)
~~~

##### **Motivation for or Use Case**

When start pure jhipster microservice application mobile gateway not working.

when i log all services : 

```java
private String getKeyFromAuthorizationServer() {

          List<String> services = discoveryClient.getServices();
          log.error("-----SERVICE------");
          for(String s: services){
              log.info("service : " + s + "");
          }

.........
```
i have got log : 

> 2016-08-18 21:38:21.868 ERROR 6346 --- [  restartedMain] .w.m.c.MicroserviceSecurityConfiguration : -----SERVICE------
2016-08-18 21:38:21.868  INFO 6346 --- [  restartedMain] .w.m.c.MicroserviceSecurityConfiguration : service : uaa

so its seems that service exist .

##### **Reproduce the error**

1. Install new Jhipster version
2. Create UAA service with cassandra
3. Create cassandra keyspace and tables
4. Create mobile-gateway
5. Copy jhipster-registry from github
6. Start jhipster-registry
7. Start uaa
8. Start mogile-gateway --- ERROR

##### **Suggest a Fix**

Something with mobile gateway configuration.
It't very weird because when i modify MicroserviceSecurityConfiguration : 

```java
private String getKeyFromAuthorizationServer() {
          List<String> services = discoveryClient.getServices();

          HttpEntity<Void> request = new HttpEntity<Void>(new HttpHeaders());
          String value = (String) this.keyUriRestTemplate
              .exchange("http://uaa/oauth/token_key", HttpMethod.GET, request, Map.class).getBody()
              .get("value");

          return value;
      }

```` 
It's working like a charm. Its seems that its lazy loading of something like that. 

So basically that one line fix issue : 

```java
List<String> services = discoveryClient.getServices();
```
 
##### **JHipster Version(s)**

"jhipsterVersion": "3.6.1",

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

~~~
{
  "generator-jhipster": {
    "jhipsterVersion": "3.6.1",
    "baseName": "turnigo_mobgateway",
    "packageName": "it.seiton.turnigo.web.mobgateway",
    "packageFolder": "it/seiton/turnigo/web/mobgateway",
    "serverPort": "8081",
    "authenticationType": "uaa",
    "uaaBaseName": "uaa",
    "hibernateCache": "no",
    "clusteredHttpSession": "hazelcast",
    "websocket": "no",
    "databaseType": "cassandra",
    "devDatabaseType": "cassandra",
    "prodDatabaseType": "cassandra",
    "searchEngine": "no",
    "buildTool": "gradle",
    "useSass": true,
    "applicationType": "gateway",
    "testFrameworks": [
      "gatling",
      "cucumber",
      "protractor"
    ],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en",
      "ca",
      "zh-cn",
      "zh-tw",
      "cs",
      "da",
      "nl",
      "fr",
      "gl",
      "de",
      "el",
      "hi",
      "hu",
      "it",
      "ja",
      "ko",
      "mr",
      "pl",
      "pt-br",
      "pt-pt",
      "ro",
      "ru",
      "sk",
      "es",
      "sv",
      "tr",
      "ta"
    ]
  }
}
~~~
Hi here is uaa yo-rc configuration
~~~

`{
  "generator-jhipster": {
    "jhipsterVersion": "3.6.1",
    "baseName": "uaa",
    "packageName": "it.seiton.turnigo.web.uaa",
    "packageFolder": "it/seiton/turnigo/web/uaa",
    "serverPort": "9999",
    "authenticationType": "uaa",
    "hibernateCache": "no",
    "databaseType": "cassandra",
    "devDatabaseType": "cassandra",
    "prodDatabaseType": "cassandra",
    "searchEngine": "no",
    "buildTool": "gradle",
    "enableTranslation": true,
    "applicationType": "uaa",
    "testFrameworks": [
      "gatling",
      "cucumber"
    ],
    "jhiPrefix": "jhi",
    "skipClient": true,
    "nativeLanguage": "en",
    "languages": [
      "en",
      "ca",
      "zh-cn",
      "zh-tw",
      "cs",
      "da",
      "nl",
      "fr",
      "gl",
      "de",
      "el",
      "hi",
      "hu",
      "it",
      "ja",
      "ko",
      "mr",
      "pl",
      "pt-br",
      "pt-pt",
      "ro",
      "ru",
      "sk",
      "es",
      "sv",
      "tr",
      "ta"
    ]
  }
}`
~~~

##### **Browsers and Operating System**

This is my screen from jhipster-registry:
<img width="1615" alt="screen shot 2016-08-18 at 21 00 42" src="https://cloud.githubusercontent.com/assets/3924636/17788696/e29d3616-6586-11e6-8af9-37336c0a9777.png">

OSX EL Capitan version 10.11.6

--- a/generators/server/templates/src/main/java/package/config/_MicroserviceSecurityConfiguration.java
+++ b/generators/server/templates/src/main/java/package/config/_MicroserviceSecurityConfiguration.java
@@ -73,6 +73,9 @@ public class MicroserviceSecurityConfiguration extends WebSecurityConfigurerAdap
 import <%=packageName%>.security.AuthoritiesConstants;
 
 import org.springframework.beans.factory.annotation.Qualifier;
+<%_ if (hibernateCache !== 'hazelcast') { _%>
+import org.springframework.cloud.client.discovery.DiscoveryClient;
+<%_ } _%>
 import org.springframework.cloud.client.loadbalancer.RestTemplateCustomizer;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
@@ -100,6 +103,11 @@ public class MicroserviceSecurityConfiguration extends ResourceServerConfigurerA
 
     @Inject
     JHipsterProperties jHipsterProperties;
+
+    <%_ if (hibernateCache !== 'hazelcast') { _%>
+    @Inject
+    DiscoveryClient discoveryClient;
+    <%_ } _%>
 
     @Override
     public void configure(HttpSecurity http) throws Exception {
@@ -144,6 +152,10 @@ public class MicroserviceSecurityConfiguration extends ResourceServerConfigurerA
     private RestTemplate keyUriRestTemplate;
 
     private String getKeyFromAuthorizationServer() {
+        <%_ if (hibernateCache !== 'hazelcast') { _%>
+        // Load available UAA servers
+        discoveryClient.getServices();
+        <%_ } _%>
         HttpEntity<Void> request = new HttpEntity<Void>(new HttpHeaders());
         return (String) this.keyUriRestTemplate
             .exchange("http://<%= uaaBaseName %>/oauth/token_key", HttpMethod.GET, request, Map.class).getBody()

--- a/generators/server/templates/src/main/java/package/config/_MicroserviceSecurityConfiguration.java
+++ b/generators/server/templates/src/main/java/package/config/_MicroserviceSecurityConfiguration.java
@@ -73,9 +73,7 @@ public class MicroserviceSecurityConfiguration extends WebSecurityConfigurerAdap
 import <%=packageName%>.security.AuthoritiesConstants;
 
 import org.springframework.beans.factory.annotation.Qualifier;
-<%_ if (hibernateCache !== 'hazelcast') { _%>
 import org.springframework.cloud.client.discovery.DiscoveryClient;
-<%_ } _%>
 import org.springframework.cloud.client.loadbalancer.RestTemplateCustomizer;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
@@ -104,10 +102,8 @@ public class MicroserviceSecurityConfiguration extends ResourceServerConfigurerA
     @Inject
     JHipsterProperties jHipsterProperties;
 
-    <%_ if (hibernateCache !== 'hazelcast') { _%>
     @Inject
     DiscoveryClient discoveryClient;
-    <%_ } _%>
 
     @Override
     public void configure(HttpSecurity http) throws Exception {
@@ -152,10 +148,8 @@ public class MicroserviceSecurityConfiguration extends ResourceServerConfigurerA
     private RestTemplate keyUriRestTemplate;
 
     private String getKeyFromAuthorizationServer() {
-        <%_ if (hibernateCache !== 'hazelcast') { _%>
         // Load available UAA servers
         discoveryClient.getServices();
-        <%_ } _%>
         HttpEntity<Void> request = new HttpEntity<Void>(new HttpHeaders());
         return (String) this.keyUriRestTemplate
             .exchange("http://<%= uaaBaseName %>/oauth/token_key", HttpMethod.GET, request, Map.class).getBody()

